src/lib/optimizer/strategy/column_pruning_rule.cpp

```diff
@@ -22,167 +23,344 @@ using namespace opossum::expression_functional;  // NOLINT
+    if (predicate->predicate_condition != PredicateCondition::Equals) continue;
+
+    const auto& left_operand = predicate->left_operand();
+    const auto& right_operand = predicate->right_operand();
+
+    join_is_unique_on_left_side |= is_unique_column(left_operand);
+    join_is_unique_on_right_side |= is_unique_column(right_operand);
+  }
+
+  // If one of the input sides is unused (i.e., its expressions are not needed in the output), check whether
+  if (!left_input_is_used && join_is_unique_on_left_side) {
+    join_node->join_mode = JoinMode::Semi;
+    const auto temp = join_node->left_input();
+    join_node->set_left_input(join_node->right_input());
+    join_node->set_right_input(temp);
+  }
+
+  if (!right_input_is_used && join_is_unique_on_right_side) {
+    join_node->join_mode = JoinMode::Semi;
+  }
 }
 
-void ColumnPruningRule::_prune_columns_in_projections(const std::shared_ptr<AbstractLQPNode>& lqp,
-                                                      const ExpressionUnorderedSet& referenced_columns) {
-  /**
-   * Prune otherwise unused columns that are forwarded by ProjectionNodes
-   */
-
-  // First collect all the ProjectionNodes. Don't prune while visiting because visit_lqp() can't deal with nodes being
-  // replaced
-  auto projection_nodes = std::vector<std::shared_ptr<ProjectionNode>>{};
-  visit_lqp(lqp, [&](auto& node) {
-    if (node->type != LQPNodeType::Projection) return LQPVisitation::VisitInputs;
-    projection_nodes.emplace_back(std::static_pointer_cast<ProjectionNode>(node));
-    return LQPVisitation::VisitInputs;
-  });
-
-  // Replace ProjectionNodes with pruned ProjectionNodes if necessary
-  for (const auto& projection_node : projection_nodes) {
-    auto referenced_projection_expressions = std::vector<std::shared_ptr<AbstractExpression>>{};
-    for (const auto& expression : projection_node->node_expressions) {
-      // We keep all non-column expressions
-      if (expression->type != ExpressionType::LQPColumn ||
-          referenced_columns.find(expression) != referenced_columns.end()) {
-        referenced_projection_expressions.emplace_back(expression);
+void prune_projection_node(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    const std::unordered_map<std::shared_ptr<AbstractLQPNode>, ExpressionUnorderedSet>& required_expressions_by_node) {
+  // Iterate over the ProjectionNode's expressions and add them to the required expressions if at least one output node
+  // requires them
+  auto projection_node = std::dynamic_pointer_cast<ProjectionNode>(node);
+
+  auto new_node_expressions = std::vector<std::shared_ptr<AbstractExpression>>{};
+  new_node_expressions.reserve(projection_node->node_expressions.size());
+
+  for (const auto& expression : projection_node->node_expressions) {
+    for (const auto& output : node->outputs()) {
+      const auto& required_expressions = required_expressions_by_node.at(output);
```

Comment:
Are you using `at()` intentionally here to get an exception in case `output` is not in `required_expressions_by_node `?

Reply:
Yes.